sycl/source/detail/graph_impl.cpp

```diff
@@ -1027,141 +1042,297 @@ exec_graph_impl::~exec_graph_impl() {
-          throw sycl::exception(
-              errc::event,
-              "Failed to enqueue event for command buffer submission");
-        }
+    bool IsLastPartition = (Partition == MPartitions.back());
+    EventImplPtr EnqueueEvent;
+
+    // We always need to request an event to use as dependency between
+    // partitions executions and between graph executions because the
+    // scheduler doesn't seem to guarantee the execution order of host-tasks
+    // without adding explicit event dependencies even when the queue is
+    // in-order.
+    constexpr bool RequestEvent = true;
+
+    if (Partition->MIsHostTask) {
+      // The event returned by a host-task is always needed to synchronize with
+      // other partitions or to be used by the sycl queue as a dependency for
+      // further commands.
+      EnqueueEvent =
+          enqueueHostTaskPartition(Partition, Queue, CGData, RequestEvent);
+    } else {
+      // The scheduler can only be skipped if the partition is a root and is not
+      // a host-task. This is because all host-tasks need to go through the
+      // scheduler and, since only the scheduler can wait on host-task events,
+      // any subsequent partitions that depend on a host-task partition also
+      // need to use the scheduler.
+      bool SkipScheduler = Partition->MPredecessors.empty() &&
+                           IsCGDataSafeForSchedulerBypass &&
+                           Partition->MRequirements.empty();
+      if (SkipScheduler) {
+        EnqueueEvent = enqueuePartitionDirectly(Partition, Queue,
+                                                CGData.MEvents, RequestEvent);
       } else {
-        std::unique_ptr<sycl::detail::CG> CommandGroup =
-            std::make_unique<sycl::detail::CGExecCommandBuffer>(
-                CommandBuffer, nullptr, std::move(CGData));
+        EnqueueEvent = enqueuePartitionWithScheduler(Partition, Queue, CGData,
+                                                     RequestEvent);
+      }
+    }
 
-        NewEvent = sycl::detail::Scheduler::getInstance().addCG(
-            std::move(CommandGroup), Queue.shared_from_this(),
-            /*EventNeeded=*/true);
+    if (!Partition->MSuccessors.empty()) {
+      // Need to keep track of the EnqueueEvent for this partition so that
+      // it can be added as a dependency to CGData when successors are executed.
+      Partition->MEvent = std::move(EnqueueEvent);
+    } else {
+      // Unified runtime guarantees the execution order of command-buffers.
+      // However, since host-tasks have been scheduled, we always need to add a
+      // dependency for the next graph execution. If we don't the next graph
+      // execution could end up with the same host-task node executing in
+      // parallel.
+      MSchedulerDependencies.push_back(EnqueueEvent);
+      if (EventNeeded) {
+        if (IsLastPartition) {
+          // If we are in the last partition copy the event to SignalEvent,
```

Comment:
```suggestion
          // If we are in the last partition move the event to SignalEvent,
```
Nit but we're not doing a copy here.